File: channel/android/app/src/main/kotlin/top/guodf/channel_example/MyChannel.kt
package top.guodf.channel_example

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.StringCodec
import java.nio.ByteBuffer

class MyChannel(private val binaryMessenger: BinaryMessenger,private val name:String, private val messageCodec: StringCodec= StringCodec.INSTANCE) {

    public fun send(msg:String){
        binaryMessenger.send(name,messageCodec.encodeMessage(msg))
    }

    fun setHandler(binaryMessageHandler: BinaryMessenger.BinaryMessageHandler?) {
        binaryMessenger.setMessageHandler(name, binaryMessageHandler);
    }
}

File: channel/lib/my_channel.dart
import 'dart:async';

import 'package:flutter/services.dart';

class MyChannel {
  final String name;
  final MessageCodec codec;
  final BinaryMessenger binaryMessenger;
  const MyChannel(this.name, this.codec,
      {this.binaryMessenger = defaultBinaryMessenger});

  Future<String> send(String msg) async {
    var data = codec.encodeMessage(msg);
    var result = await defaultBinaryMessenger.send(name, data);
    return codec.decodeMessage(result);
  }

  Stream<String> eventStream(String msg) {
    var controller = StreamController<String>.broadcast();
    defaultBinaryMessenger.setMessageHandler(name, (data) {
      if (data == null) controller.close();
      var value = codec.decodeMessage(data);
      controller.add(value);
      return Future.value(data);
    });
    Future(() async {
      await for (var _ in controller.stream) {
        send(msg);
      }
    });
     send(msg);
    return controller.stream;
  }
}
